src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -352,10 +343,22 @@ void ShenandoahOldHeuristics::prepare_for_old_collections() {
         size_t region_count = heap->trash_humongous_region_at(region);
         log_debug(gc)("Trashed " SIZE_FORMAT " regions for humongous object.", region_count);
       }
-    } else if (region->is_trash()) {
-      // Count humongous objects made into trash here.
-      immediate_regions++;
-      immediate_garbage += garbage;
+    } else if (region->is_regular() || region->is_pinned()) {
+      if (!region->has_live()) {
```

Comment:
Here, I'm not sure why we ask "|| region->is_pinned()".  We don't want humongous regions that are pinned.  And we do want regular regions whether or not they are pinned.  Also, I think it better to handle the case of !region->has_live() at line 327 above.

Reply:
I agree, it's confusing to have `is_pinned` mask the underlying state of the region.